### ISSUE TITLE
DINOv3-7B Crashes ComfyUI

### ISSUE THREAD

**Issue by ialhabbal** [HM-RunningHub/ComfyUI_RH_ZImageI2L#3]
DINOv3-7B Crashes ComfyUI.

**Comment by masslevel** [comment]
It sometimes randomly crashes ComfyUI. When ComfyUI was restarted it works again - without making any changes to the pipeline. I also noticed when the input images are above 1 - 2 megapixel, it crashes as well. I now resize my input images between 1 - 2 megapixel and it did fix at least this crash condition for me.

**Comment by linwenzi19830415-debug** [comment]
我也是，加载完模型就崩溃了

**Comment by llyyfree** [comment]
+1

**Comment by llyyfree** [comment]
我找到了出错的位置，把nodes.py的self.vram_config_disk_offload设置为={"offload_dtype":"disk","offload_device":"disk","onload_dtype": "disk","onload_device": "disk", ...下面的不用改,,,} 就行。我没有深究其背后原理，下面是我跟AI的问答，我也不确定它说得对不对，有兴趣的可以看一下：

崩溃关键位置：
def load_model(model_class, path, config=None, torch_dtype=torch.bfloat16, device="cpu", state_dict_converter=None, use_disk_map=False, module_map=None, vram_config=None, vram_limit=None, state_dict=None):
    print(f"load_model:::{model_class} {path} {config} {torch_dtype} {device} {state_dict_converter} {use_disk_map} {module_map} {vram_config} {vram_limit} {state_dict}")
    config = {} if config is None else config
    with ContextManagers(get_init_context(torch_dtype=torch_dtype, device=device)):
        model = model_class(**config)
    # What is `module_map`?
    # This is a module mapping table for VRAM management.
    if module_map is not None:
        print(f"load_model 1111")
        devices = [vram_config["offload_device"], vram_config["onload_device"], vram_config["preparing_device"], vram_config["computation_device"]]
        device = [d for d in devices if d != "disk"][0]
        dtypes = [vram_config["offload_dtype"], vram_config["onload_dtype"], vram_config["preparing_dtype"], vram_config["computation_dtype"]]
        dtype = [d for d in dtypes if d != "disk"][0]
        if vram_config["offload_device"] != "disk":
            print(f"load_model 1111 111")
            if state_dict is None: state_dict = DiskMap(path, device, torch_dtype=dtype)
            if state_dict_converter is not None:
                print(f"load_model 1111 222")
                state_dict = state_dict_converter(state_dict)
            else:
                state_dict = {i: state_dict[i] for i in state_dict}
            print(f"load_model 1111 333")
            model.load_state_dict(state_dict, assign=True)
            model = enable_vram_management(model, module_map, vram_config=vram_config, disk_map=None, vram_limit=vram_limit)
        else:
            print(f"load_model 1111 444")
            disk_map = DiskMap(path, device, state_dict_converter=state_dict_converter)
            model = enable_vram_management(model, module_map, vram_config=vram_config, disk_map=disk_map, vram_limit=vram_limit)

崩溃日志：
Downloading Model from https://www.modelscope.cn to directory: F:\ruanjian\ComfyUI_windows_portable\ComfyUI\models\DiffSynth-Studio\General-Image-Encoders
Loading models from: "F:\\ruanjian\\ComfyUI_windows_portable\\ComfyUI\\models\\DiffSynth-Studio/General-Image-Encoders\\DINOv3-7B/model.safetensors"
load_model:::<class 'diffsynth.models.dinov3_image_encoder.DINOv3ImageEncoder'> F:\ruanjian\ComfyUI_windows_portable\ComfyUI\models\DiffSynth-Studio/General-Image-Encoders\DINOv3-7B/model.safetensors {} torch.bfloat16 cuda None True {<class 'transformers.models.dinov3_vit.modeling_dinov3_vit.DINOv3ViTLayerScale'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'transformers.models.dinov3_vit.modeling_dinov3_vit.DINOv3ViTRopePositionEmbedding'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'transformers.models.dinov3_vit.modeling_dinov3_vit.DINOv3ViTEmbeddings'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'torch.nn.modules.conv.Conv2d'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'torch.nn.modules.normalization.LayerNorm'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'torch.nn.modules.linear.Linear'>: <class 'diffsynth.core.vram.layers.AutoWrappedLinear'>} {'offload_device': 'cpu', 'offload_dtype': torch.bfloat16, 'onload_device': 'cuda', 'onload_dtype': torch.bfloat16, 'preparing_device': 'cuda', 'preparing_dtype': torch.bfloat16, 'computation_device': 'cuda', 'computation_dtype': torch.bfloat16} 13.99951171875 None
load_model 1111
load_model 1111 111
Press any key to continue . . .

它的配置：
self.vram_config_disk_offload = {
            "offload_dtype": torch.bfloat16,
            "offload_device": "cpu",
            "onload_dtype": torch.bfloat16,
            "onload_device": "cuda",
            "preparing_dtype": torch.bfloat16,
            "preparing_device": "cuda",
            "computation_dtype": torch.bfloat16,
            "computation_device": "cuda",
        }

另一个插件成功运行的日志：
Downloading Model from https://www.modelscope.cn to directory: F:\ruanjian\ComfyUI_windows_portable\ComfyUI\models\DiffSynth-Studio\General-Image-Encoders
Loading models from: "F:\\ruanjian\\ComfyUI_windows_portable\\ComfyUI\\models\\DiffSynth-Studio/General-Image-Encoders\\DINOv3-7B/model.safetensors"
load_model:::<class 'diffsynth.models.dinov3_image_encoder.DINOv3ImageEncoder'> F:\ruanjian\ComfyUI_windows_portable\ComfyUI\models\DiffSynth-Studio/General-Image-Encoders\DINOv3-7B/model.safetensors {} torch.bfloat16 cuda None True {<class 'transformers.models.dinov3_vit.modeling_dinov3_vit.DINOv3ViTLayerScale'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'transformers.models.dinov3_vit.modeling_dinov3_vit.DINOv3ViTRopePositionEmbedding'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'transformers.models.dinov3_vit.modeling_dinov3_vit.DINOv3ViTEmbeddings'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'torch.nn.modules.conv.Conv2d'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'torch.nn.modules.normalization.LayerNorm'>: <class 'diffsynth.core.vram.layers.AutoWrappedModule'>, <class 'torch.nn.modules.linear.Linear'>: <class 'diffsynth.core.vram.layers.AutoWrappedLinear'>} {'offload_device': 'disk', 'offload_dtype': 'disk', 'onload_device': 'disk', 'onload_dtype': 'disk', 'preparing_device': 'cuda', 'preparing_dtype': torch.bfloat16, 'computation_device': 'cuda', 'computation_dtype': torch.bfloat16} 13.99951171875 None
load_model 1111
load_model 1111 444

它的配置：
self.vram_config_disk_offload = {
            "offload_dtype": "disk",
            "offload_device": "disk",
            "onload_dtype": "disk",
            "onload_device": "disk",
            "preparing_dtype": torch.bfloat16,
            "preparing_device": "cuda",
            "computation_dtype": torch.bfloat16,
            "computation_device": "cuda",
        }

我找到了出错的准确位置，只需要把配置改了就行，但我不明白背后原理，麻烦解释一下。

Gemini 说
真相大白了！你抓到的这几行日志和代码对比，精准地暴露了 diffsynth 库在处理大模型加载时的一个**“逻辑陷阱”**。
简单来说：当你配置为 cpu 时，代码在试图**“一口气吞掉”整个 14GB 的模型；而配置为 disk 时，它在“分口细品”**。
以下是背后的技术原理分析：
1. 崩溃点：那个致命的“字典推导式”
观察崩溃的代码段（offload_device != "disk" 分支）：
Python
if state_dict is None: state_dict = DiskMap(path, device, torch_dtype=dtype)
if state_dict_converter is not None:
    # ...
else:
    state_dict = {i: state_dict[i] for i in state_dict}  # <--- 就在这里崩溃
原理：
DiskMap 本身是无害的： 它只是建立了一个磁盘映射（mmap），就像一张地图，告诉你数据在硬盘的什么位置，此时并不占用真正的 RAM。
命的操作： {i: state_dict[i] for i in state_dict} 这一行。这行代码在遍历 DiskMap，并强迫操作系统把硬盘上 DINOv3-7B 的每一个权重（约 14GB）立刻、全部加载到 Python 的字典对象里。
为什么 48G 内存也会崩： 1.  瞬间压力： 这个循环运行速度极快，系统会瞬间产生巨大的 I/O 和内存分配请求。
2.  双倍占用： 当这个字典创建完后，紧接着执行 model.load_state_dict(state_dict, assign=True)。此时，内存中存在一份 14GB 的字典和一份 14GB 的模型实例。在 assign 完成前的瞬间，你至少需要 14×2=28 GB 的连续物理内存。
3.  Python 的开销： Python 字典存储成千上万个小张量（Tensors）会有额外的对象开销。虽然你虚拟内存大，但这种短时间内的剧烈内存申请容易触发操作系统的保护机制（或底层 C++ 库的内存溢出），直接杀掉进程。
2. 为什么改成 disk 就成功了？
观察成功运行的代码段（else 分支）：
Python
disk_map = DiskMap(path, device, state_dict_converter=state_dict_converter)
model = enable_vram_management(model, module_map, vram_config=vram_config, disk_map=disk_map, vram_limit=vram_limit)
原理：
跳过了“死亡循环”： 它没有去创建那个庞大的 Python 字典，而是直接把 DiskMap（那个地图）传给了 enable_vram_management（显存管理器）。
按需加载（Lazy Loading）： 显存管理器非常聪明。它不会一次性读入 14GB，而是等到模型运行到某一层时，才去磁盘上抓取那一层需要的权重。
内存极其平稳： 这种方式下，内存占用始终维持在极低的水平，因为大部分权重都安安静静地躺在硬盘（磁盘映射）里。
3. 为什么另一个插件 Qwen 没事？
既然两个插件共用底层，Qwen 能跑通只有一种可能：它的开发者在插件初始化时，默认就把 offload_device 设为了 "disk"。
而你崩溃的那个插件 ZImage，它的开发者可能默认将其设为了 "cpu"。虽然 cpu 理论上比 disk 快一点点（因为减少了磁盘寻址），但它极其吃内存，且在处理 DINOv3 这种 7B 级别的模型时，直接撞上了 Python 处理大字典的效率瓶颈。
总结与建议
结论： 崩溃不是因为你硬件不够，而是 diffsynth 在 offload_device="cpu" 模式下的实现方式太暴力，试图在内存中瞬间“克隆”整个大模型，导致了进程自杀。
建议：
永久改为 disk： 对于 7B 这种规模的模型，永远推荐使用 offload_device: "disk"。在 NVME 固态硬盘普及的今天，磁盘加载和 CPU 加载的性能差异微乎其微，但稳定性天差地别。
修改插件配置： 你可以去 ComfyUI_RH_ZImageI2L 的源代码里搜索 offload_device，把默认值从 "cpu" 改成 "disk"，这样以后就不用手动改了。
既然现在问题解决了，你需要我帮你写一段脚本，一次性扫描并修改该插件所有节点的默认显存配置吗？
